### ISSUE TITLE
Upgrading v4 to v5 panic: Listener was never inserted into the list

### ISSUE THREAD

**Issue by jbr** [smol-rs/event-listener#124]
I'm unsure if I'm holding the new interface wrong somehow or if I've encountered a bug.

Given [this diff](https://github.com/trillium-rs/trillium/pull/595/files) that attempts to track the event-listener interface changes, I'm encountering a deterministic panic like:
```
event-listener-5.2.0/src/lib.rs:1251:36:
listener was never inserted into the list
```

and as far as I can tell the code reduces to

```rust
let n = AtomicUsize::from(1);
let event = Event::new();
let listener = event.listen();
loop {
    if 0 == n.load(Ordering::SeqCst) {
        return Poll::Ready(());
    };
    ready!(Pin::new(&mut self.listener).poll(cx));
}
```

Do I need to replace the listener every time I poll it? Am I holding this wrong? Is this a bug?

**Comment by notgull** [comment]
It seems you are using the future like this:

```rust
let mut listener = event.listen();
loop {
    if 0 == n.load(Ordering::SeqCst) {
        return Poll::Ready(());
    };
    ready!(Pin::new(&mut listener).poll(cx));
} 
```

It should be used like this:

```rust
let mut listener = None;
loop {
    if 0 == n.load(Ordering::SeqCst) {
        return Poll::Ready(());
    };
    if let Some(listener) = listener.as_mut() {
        ready!(Pin::new(&mut listener).poll(cx));
        listener = None;
    } else {
        listener = Some(event.listen());
    }
}
```

Polling the `EventListener` after it has returned `Ready` will panic, and further polls should use a completely new `EventListener` instead.

The v4 API kept track of this state inside of the `EventListener,` but this [caused confusion](https://github.com/smol-rs/event-listener/issues/89) and was removed in v5.

**Comment by notgull** [comment]
If you think this could be made clearer in documentation I would accept a PR.

**Comment by jbr** [comment]
Would it make sense to just always return ready once it has ever been ready instead of panicking?

**Comment by notgull** [comment]
Rust's documentation seems to imply that [a panic is the preferred option](https://docs.rs/futures/latest/futures/future/trait.Future.html#panics) in this case.

**Comment by jbr** [comment]
It looks like EventListener holds an InnerListener that contains a reference to the Event. Would it work to have a configurable option that allows the listener to reregister itself on completion? It feels awkward to require user code to `match Pin::new(&mut listener).poll(cx) { Poll::Ready(()) => { std::mem::replace(listener, event.listen(); ... }, ... }`

**Comment by notgull** [comment]
It's a bad idea to have the `EventListener` automatically insert itself into the list. Generally, the flow should be:

- Check your condition.
- Register the listener into the list.
- Check your condition, ideally with stronger ordering.
- Wait on the listener.

If you register the listener into the list without actively polling an event, it can lead to strange behavior. E.g. your future received a notification meant for another task
